include/deal.II/base/polynomials_hermite.h

```diff
@@ -0,0 +1,195 @@
+     * obey the following conditions:
+     *
+     * @f{align*}{
+     * &\begin{matrix}
+     *   \left. \frac{d^{i}}{dy^{i}} H(y) \right\vert_{y=0}
+     *          = i! 4^{i} \delta_{i, index}, \hfill
+     *          &\qquad \hfill 0 \leq i \leq regularity, \\
+     *   \left. \frac{d^{i}}{dy^{i}} H(y) \right\vert_{y=1}
+     *          = 0, \hfill &\qquad \hfill 0 \leq i \leq regularity,
+     * \end{matrix} \qquad 0 \leq index \leq regularity, \\
+     * &\begin{matrix}
+     *  \left. \frac{d^{i}}{dy^{i}} H(y) \right\vert_{y=0}
+     *          = 0, \hfill &\qquad \hfill 0 \leq i \leq regularity, \\
+     * \left. \frac{d^{i}}{dy^{i}} H(y) \right\vert_{y=1}
+     *          = i! 4^{i} \delta_{i, index-(regularity+1)}, \hfill
+     *          &\qquad \hfill 0 \leq i \leq regularity,
+     *\end{matrix} \qquad regularity+1 \leq index
+     *          \leq 2 \times regularity + 1,
+     * @f}
+     *
+     * where $\delta_{i,j}$ is equal to $1$ whenever $i=j$,
+     * and equal to $0$ otherwise.
+     *
+     * @param regularity
+     * @param index
+     */
+    HermitePoly(const unsigned int regularity, const unsigned int index)
+      : Polynomial<double>(hermite_poly_coeffs(regularity, index))
+      , degree(2 * regularity + 1)
+      , regularity(regularity)
+      , side_index(index % (regularity + 1))
+      , side((index > regularity) ? 1 : 0){};
+
+
+
+    /**
+     * This function is used internally to generate a vector of Polynomial
+     * objects representing a complete basis of degree
+     * $2 \times$ @p regularity $+1$ on the reference interval $[0,1]$.
+     *
+     * @param regularity
+     */
+    static std::vector<Polynomial<double>>
+    generate_complete_basis(const unsigned int regularity);
+
+
+
+    /**
+     * Function that returns a matrix $\mathbf{B}$
+     * of size $($ @p regularity $+1) \times ($ @p regularity $+1)$
+     *containing derivatives of Bernstein polynomials of the
+     * form
+     *
+     * @f{align*}{
+     *  z_{k}(y) = y^{k} (1-y)^{regularity+1},
+     *          \qquad 0 \leq k \leq regularity.
+     * @f}
+     *
+     * We define $j! B_jk$ to be the $j^{th}$ derivative of
+     * polynomial $z_{k}(y)$ at $y=0$ for $0 \leq j \leq $ @p regularity.
+     * This is useful both for computing the basis and assigning
+     * weights to local degrees of freedom at hanging nodes. The
+     * rescaling is to improve performance of solution, by ensuring
+     * the main diagonal consists entirely of $1$s.
+     *
+     * @param regularity
+     */
+    static FullMatrix<double>
+    hermite_to_bernstein_matrix(const unsigned int regularity);
+
+
+    /**
+     * This function generates a vector of coefficients that can be
+     * used by the existing Polynomial class constructor to form
+     * the required objects.
+     */
+    static std::vector<double>
+    hermite_poly_coeffs(const unsigned int regularity,
+                        const unsigned int index);

```

Comment:
I started investigating the possibility of implementing partial grid refinement with Hermite elements, and the function hermite_to_bernstein_matrix will be needed then so I would prefer to leave it as publicly available for now. I have moved hermite_poly_coeffs into an internal namespace since it wasn't needed in the header file when the constructor is defined in the source code file.